### ISSUE TITLE
Library incompatible with psr/cache ^3.0

### ISSUE THREAD

**Issue by Geolim4** [googleapis/google-auth-library-php#363]
Hello,

I'm running a project ([Phpfastcache#v9](https://github.com/PHPSocialNetwork/phpfastcache/tree/v9)) that requires `psr/cache ^3.0`, which conflict with your constraints: `^1.0|^2.0`

So actually I cannot make use of any Google product at all because of this very "central" constraint.

Do you plan to make it compatible with `psr/cache ^3.0` ?

It's basically the same interface as previous but with PHP-8 type hint.

This should not break compatibility at all to support it since `psr/cache ^3.0`  already constraints on `php ^8.0` and the interface remains the same.

Thanks.

#### Environment details

  - OS: Win 10 x64
  - PHP version: 8.0
  - Package name and version: latest

**Comment by Geolim4** [comment]
@bshaffer this issue will become more and more blocking over the months coming with the PHP-8 popularity growing up :)

**Comment by bshaffer** [comment]
Hi @Geolim4 ! yes you're right, and it's a concern. But correct me if I'm wrong, every project supporting `3.0` _should_ be supporting `2.0`, and so as long as all the projects which support `1.0` also support `2.0`, the problem should be resolved.

Upgrading to `3.0` would break BC and require a major version bump, which we may do if this issue becomes an issue for other auth library consumers.

**Comment by Geolim4** [comment]
> Upgrading to `3.0` would break BC and require a major version bump, which we may do if this issue becomes an issue for other auth library consumers.

If you add the `|| ^3.0` the library _should not_ have a BC break (or I don't see it :P):

If you project make use of `google/auth` as a dependecy:

- Without `psr/cache ^3.0` support (Current behavior)
  - You use php < 8.0:  no problem, composer will stay on a PHP-compatible version of `psr/cache`: `1.0.1`
  - You use php >= 8.0: no problem, composer will select a PHP-compatible version of `psr/cache`: `1.0.1` or `2.0.0`
- With `psr/cache ^3.0` support (new behavior)
  - You use php < 8.0: , no problem, composer will stay on a PHP-compatible version of `psr/cache`: `1.0.1` or `2.0.0` (`3.0.0`  is indeed incompatible, but not blocking for Composer)
  - You use php >= 8.0: Your project is already compatible and working with `google/auth`, adding a new compatibility will only allow composer to make use of PHP-8 benefits for `psr/cache`: Upgrade available to `psr/cache ^3.0` without BC: The interface signatures are not incompatible at project-level (only PHP-level incompatible, which is here irrelevant), the interfaces signatures remain the same, the project is still working and has benefits of PHP-8 enforced types.
 


Did I missed something ? Don't hesitate to be technical and precise :D

**Comment by Geolim4** [comment]
Hello @bshaffer, I plan to add Firestore support on Phpfastcache (v9) which will be released by half-October, do you have any ETA to fix this "issue" ?

Thanks you :)

**Comment by bshaffer** [comment]
@Geolim4 I am not sure how you think adding `3.0` _won't_ break compatibility. The interfaces are not compatible with our classes, and if we upgrade our classes we will break backwards compatibility with earlier libraries.

**Comment by Geolim4** [comment]
> 
> 
> @Geolim4 I am not sure how you think adding `3.0` _won't_ break compatibility. The interfaces are not compatible with our classes, and if we upgrade our classes we will break backwards compatibility with earlier libraries.

The interfaces haven't changed, only the signature changed, and just became more typed if you have PHP-8 installed.
If your project don't run on PHP8, composer will downgrade to 1.0.1.

The only thing to check is to make sure that you've passed the good type to the PSR interface before even if it was not strictly typed.

Aside of that, I'm not sure that you encounter any BC break issue.

**Comment by bshaffer** [comment]
Please see https://github.com/googleapis/google-auth-library-php/pull/364#issue-742424942 for why we cannot support `psr/cache:3` without major considerations.

Also, please try getting the repositories which require `psr/cache:^3` to instead require `psr/cache:^2||^3`, as this would solve the problem for everyone.

**Comment by Geolim4** [comment]
> Also, please try getting the repositories which require `psr/cache:^3` to instead require `psr/cache:^2||^3`, as this would solve the problem for everyone.

I tried, but the dependency still refused to be installed by Composer :/

**Comment by bshaffer** [comment]
> There's another library involved in this issue: google/grpc-gcp

Thanks for pointing this out! I've submitted a PR to add support for both v2 and v3 there.

**Comment by Geolim4** [comment]
I just noticed that there's another dependency lock:

```
  Problem 1
    - google/grpc-gcp[0.1.0, ..., 0.1.5] require psr/cache ^1.0.1 -> found psr/cache[1.0.1] but it conflicts with your root composer.json require (^2.0||^3.0).
    - google/cloud-firestore v1.20.0 requires google/gax ^1.1 -> satisfiable by google/gax[1.1.0, ..., v1.9.0].
    - google/gax[1.1.0, ..., v1.9.0] require google/grpc-gcp ^0.1.0 -> satisfiable by google/grpc-gcp[0.1.0, ..., 0.1.5].
    - Root composer.json requires google/cloud-firestore ^1.20 -> satisfiable by google/cloud-firestore[v1.20.0].

```

As you can see `google/gax` [locked ](https://github.com/googleapis/gax-php/blob/master/composer.json) `google/grpc-gcp` to `^0.1.0` which make impossible the upgrade from `0.1.5` to `0.2.0` preventing by the way `google/cloud-firestore` to benefit a partial fix with `psr/cache`  `2.0`, awaiting the current merge request to add an eventual full support for `psr/cache`  `3.0`.

**Comment by bshaffer** [comment]
@Geolim4 a new version of `google/grpc-gcp` [has been released](https://github.com/GoogleCloudPlatform/grpc-gcp-php/releases/tag/v0.2.0). Try updating your dependencies with `composer update` and see if you can now install `psr/cache:v2`. 

Unfortunately, as long as you are using phpfastcache, you will not be able to resolve your dependencies until https://github.com/PHPSocialNetwork/phpfastcache/pull/833 is merged and tagged.

**Comment by bshaffer** [comment]
Closing this issue as we now support the compatibility layer `psr/cache:v2`, which is all that we plan on doing at this time.

**Comment by Geolim4** [comment]
> @Geolim4 a new version of `google/grpc-gcp` [has been released](https://github.com/GoogleCloudPlatform/grpc-gcp-php/releases/tag/v0.2.0). Try updating your dependencies with `composer update` and see if you can now install `psr/cache:v2`.
> 
> Unfortunately, as long as you are using phpfastcache, you will not be able to resolve your dependencies until [PHPSocialNetwork/phpfastcache#833](https://github.com/PHPSocialNetwork/phpfastcache/pull/833) is merged and tagged.

I'm working on it, locally, but there's still a last blocking bug that I described above ^ :D
https://github.com/googleapis/google-auth-library-php/issues/363#issuecomment-928429475

**Comment by bshaffer** [comment]
@Geolim4 this has been resolved in https://github.com/googleapis/gax-php/pull/341

**Comment by Geolim4** [comment]
My bad, just saw that [you also updated](https://github.com/googleapis/gax-php/pull/341) `google/gax`, everything is fine now, thanks 🎉

**Comment by nyroDev** [comment]
Hello,
is there any plan to update or provide a new version compatible with psr/cache^3 ?

Thanks in advance.

**Comment by bshaffer** [comment]
This is now released in [google/auth:v1.20.0](https://github.com/googleapis/google-auth-library-php/releases/tag/v1.20.0)
